### ISSUE TITLE
Completion of error handling

### ISSUE THREAD

**Issue by elfring** [jpbruyere/vkvg#16]
Would you like to add more error handling for return values from functions like the following?
- [malloc](http://pubs.opengroup.org/onlinepubs/9699919799/functions/malloc.html "Memory allocation") ⇒ [vkvg_create](https://github.com/jpbruyere/vkvg/blob/b983c8bce31bc86955a748e2f1304243f6e39cb9/src/vkvg_context.c#L34)
- [realloc](http://pubs.opengroup.org/onlinepubs/9699919799/functions/realloc.html "Memory reallocation") ⇒ [_increase_font_tex_array](https://github.com/jpbruyere/vkvg/blob/b983c8bce31bc86955a748e2f1304243f6e39cb9/src/vkvg_fonts.c#L106)

**Comment by jpbruyere** [comment]
I've not yet implemented any exception handling, but if you have a clear idea of what you want, feel free to submit a pull request.

**Comment by elfring** [comment]
I suggest to avoid [ignorance of return values](https://cwe.mitre.org/data/definitions/252.html "Checking more return values") a bit more.
Would you like to detect every error situation as early as possible?

**Comment by jpbruyere** [comment]
You are right that error handling is a key part, but I feel more comfortable with the idea of implementing a single error handling pattern as a whole in a dedicated step of the development. I guess the advantage is to keep the code as simple (and readable) as possible in the early steps of creation. And as long  as the software is run in a debug environment, error handling isn't critical. But any idea and help is welcome.

**Comment by elfring** [comment]
How do you think about to improve [static source code analysis](https://en.wikipedia.org/wiki/List_of_tools_for_static_code_analysis#C.2C_C.2B.2B "Tools for code analysis") also for this software?

-----

* Are you interested to apply [aspect-oriented software development](https://aspectc.org/ "Software development with AspectC++?")?
* Would you like to encapsulate error detection and corresponding exception handling as [a reusable aspect](https://en.wikipedia.org/wiki/Aspect_%28computer_programming%29 "Combination of a pointcut with advice") in your software?

**Comment by jpbruyere** [comment]
Static source code analyzers are handy tools, but using them too early could limit creativity and intuition.  Also being rigorous by yourself instead of being driven by a soft is a must. Letting a software making lots of tiny changes in your draft could have bad consequence.
For error handling, I will try to make them non-blocking where possible, but producing a reusable aspect is not really in my plans.

**Comment by elfring** [comment]
Can higher level development tools help to improve also this software implementation further?

**Comment by jpbruyere** [comment]
You think to a particular tool?

**Comment by elfring** [comment]
Yes. - There are development tools available which could help more with the desired completion of exception handling.
:crystal_ball: How do you think about to extend the software tool box?

**Comment by jpbruyere** [comment]
tell me more...which soft?

**Comment by elfring** [comment]
Examples:
* Programming language “C++”
* Semantic patch language (Coccinelle)
* AspectC++

**Comment by jpbruyere** [comment]
If you have good experience with those tools, I'm open to any pull request that could improve vkvg.

**Comment by elfring** [comment]
I know these tools to some degree.
:crystal_ball: How will their application acceptance evolved by other software designers, system testers and developers?

**Comment by elfring** [comment]
* How do you think about to work with functions which can throw C++ exceptions?
* Would you like to [integrate the tool “AspectC++”](https://aspectc.org/ "Software development with AspectC++") into [your software build process](https://github.com/jpbruyere/vkvg/blob/a35493023e00e276df7df29eaaeeeaa185f45c98/CMakeLists.txt#L3 "Build specification")?

**Comment by jpbruyere** [comment]
No, vkvg is a c library.

**Comment by elfring** [comment]
Can functions of a C library call (member) functions of a class library for this software?

**Comment by jpbruyere** [comment]
I've started implementing unit testing with google_test in my dev branch (https://github.com/jpbruyere/vkvg/blob/VkvgCreateInfo/), expecting error handling to be extensively improved and tested. I'll keep track of error handling strategy for vkvg in https://github.com/jpbruyere/vkvg/blob/VkvgCreateInfo/doc/design/error_handling.md
